rialto_airflow/harvest/openalex.py

```diff
@@ -1,178 +1,101 @@
-import csv
+import json
 import logging
 import os
-import pickle
 import time
+from pathlib import Path
 
-from more_itertools import batched
-from pyalex import Authors, Works, config, api
+from pyalex import Authors, Works, config
+from typing import Generator
+from sqlalchemy.dialects.postgresql import insert
 
-from rialto_airflow.utils import invert_dict, normalize_doi, normalize_orcid
+from rialto_airflow.database import (
+    Author,
+    Publication,
+    get_session,
+    pub_author_association,
+)
+from rialto_airflow.snapshot import Snapshot
+from rialto_airflow.utils import normalize_doi
 
 config.email = os.environ.get("AIRFLOW_VAR_OPENALEX_EMAIL")
 config.max_retries = 5
 config.retry_backoff_factor = 0.1
 config.retry_http_codes = [429, 500, 503]
 
 
-def doi_orcids_pickle(authors_csv, pickle_file, limit=None):
+def harvest(snapshot: Snapshot, limit=None) -> Path:
     """
-    Pass in the Authors CSV and generate a DOI -> ORCID mapping as a pickle file.
+    Walk through all the Author ORCIDs and generate publications for them.
     """
-    with open(authors_csv, "r") as csv_input:
-        orcid_dois = {}
-        count = 0
-        for row in csv.DictReader(csv_input):
-            orcid = normalize_orcid(row["orcidid"])
-            if orcid:
-                count += 1
-                orcid_dois[orcid] = list(dois_from_orcid(orcid))
-                if limit is not None and count > limit:
+    jsonl_file = snapshot.path / "openalex.jsonl"
+    count = 0
+    stop = False
+
+    with jsonl_file.open("w") as jsonl_output:
+        with get_session(snapshot.database_name).begin() as select_session:
```

Comment:
Is it right to create two separate sqlalchemy sessions if I want to do inserts to publications while selecting from the authors table?

Reply:
Thanks for pointing out our database connections are currently pulling all results back into the client. In this case it's a small number of rows (7000?) so it's not a big deal. But this will be important when we are doing the "filling in" work where we iterate through all the publications and add look for missing metadata.

Unfortunately once the Session goes out of scope the ORM Author objects no longer can be used because they are in a "detached" state. Instead I could query the `cap_profile_id` and `author_id` as a list of dictionaries. But it's nice having the Author object I think, so I'm going to leave it the way it is, at least for now.